### PR TITLE
Create SR module without native implementation

### PR DESCRIPTION
**Pull request by louiszawadzki**
### What does this PR do?

This creates a module for SR with no actual logic on the native side.
The configuration on the JS side is done, no more thing should be needed here after.

The package cannot be published to npm, even by accident. 

### Additional Notes

A coming PR will enable SR on both Android and iOS and add test + linting on native code in Bitrise.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)
- [ ] If this PR is auto-generated, please make sure also to manually update the code related to the change

### CODE REVIEW

File: packages/react-native-session-replay/android/build.gradle [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+def getExtOrIntegerDefault(name) {
+  return rootProject.ext.has(name) ? rootProject.ext.get(name) : (project.properties['DatadogSDKReactNativeSessionReplay_' + name]).toInteger()
+}
+
+android {
+  compileSdkVersion getExtOrIntegerDefault('compileSdkVersion')
+  buildToolsVersion getExtOrDefault('buildToolsVersion')
+  def agpVersion = com.android.Version.ANDROID_GRADLE_PLUGIN_VERSION
+  if (agpVersion.tokenize('.')[0].toInteger() >= 7) {
+    namespace = "com.datadog.sessionreplay.reactnative"
```
Comment (louiszawadzki) [inline packages/react-native-session-replay/android/build.gradle]: Let me know if you think we can come up with a better package name (the core package name is `com.datadog.reactnative`

File: packages/react-native-session-replay/package.json [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    ],
+    "author": "Datadog (https://github.com/DataDog)",
+    "homepage": "https://github.com/DataDog/dd-sdk-reactnative#readme",
+    "repository": "https://github.com/DataDog/dd-sdk-reactnative",
+    "bugs": {
+        "url": "https://github.com/DataDog/dd-sdk-reactnative/issues"
+    },
+    "license": "Apache-2.0",
+    "main": "lib/commonjs/index",
+    "private": true,
```
Comment (louiszawadzki) [inline packages/react-native-session-replay/package.json:19]: this is the line that prevents the package from being published

**Review by buraizu — DISMISSED** [review]
Approving the `packages/react-native-session-replay/README.md` update, but just wanted to ask whether it's necessary to add a page for something that's not supported yet. Is this something that's expected later? 

I might be misunderstanding something or not have all the relevant context, but flagging just in case.

File: packages/react-native-session-replay/android/.settings/org.eclipse.buildship.core.prefs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+arguments=
+auto.sync=false
+build.scans.enabled=false
+connection.gradle.distribution=GRADLE_DISTRIBUTION(VERSION(6.0))
+connection.project.dir=
+eclipse.preferences.version=1
+gradle.user.home=
+java.home=/Library/Java/JavaVirtualMachines/jdk1.8.0_144.jdk/Contents/Home
```
Comment (0xnm) [inline packages/react-native-session-replay/android/.settings/org.eclipse.buildship.core.prefs]: is this file needed? it contains very old references to Gradle 6.0, which is not used for sure, and Java 11 is needed at least instead of the Java 1.8 referenced here.

File: packages/react-native-session-replay/android/build.gradle (hunk 1 above)
Comment (0xnm) [reply packages/react-native-session-replay/android/build.gradle]: yes, let's flip the parts and call it `com.datadog.reactnative.sessionreplay`, this will align with the naming on the native side, which is `com.datadog.android.sessionreplay`

File: packages/react-native-session-replay/android/build.gradle [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+  config = files("$projectDir/detekt.yml")
+  reports {
+    xml {
+      enabled = true
+      destination = file("build/reports/detekt.xml")
+    }
+  }
+}
+
+// This is needed to make unmock work. Choreographer class was loaded from android.jar of Android SDK
```
Comment (0xnm) [inline packages/react-native-session-replay/android/build.gradle]: There is no more `Choreographer` class usage in Android SDK, is this configuration lambda needed?

File: packages/react-native-session-replay/android/gradle.properties [hunk 5]
```diff
@@ -0,0 +1,4 @@
+DatadogSDKReactNativeSessionReplay_kotlinVersion=1.7.21
+DatadogSDKReactNativeSessionReplay_compileSdkVersion=31
+DatadogSDKReactNativeSessionReplay_buildToolsVersion=31.0.0
+DatadogSDKReactNativeSessionReplay_targetSdkVersion=31
```
Comment (0xnm) [inline packages/react-native-session-replay/android/gradle.properties]: minor question: given that the current Android API is 34, should we target it?

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/sessionreplay/reactnative/DdSessionReplayImplementation.kt [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * The entry point to use Datadog's Session Replay feature.
+ */
+class DdSessionReplayImplementation() {
+    /**
+     * Enable session replay and start recording session.
+     * @param replaySampleRate The sample rate applied for session replay.
+     * @param defaultPrivacyLevel The privacy level used for replay.
+     */
+    fun enable(replaySampleRate: Double, defaultPrivacyLevel: String, promise: Promise) {
```
Comment (0xnm) [inline packages/react-native-session-replay/android/src/main/kotlin/com/datadog/sessionreplay/reactnative/DdSessionReplayImplementation.kt]: seems like this method doesn't actually enable anything? is it on purpose?

File: packages/react-native-session-replay/android/src/newarch/kotlin/com/datadog/sessionreplay/reactnative/DdSessionReplay.kt [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import com.facebook.react.bridge.ReactMethod
+
+/**
+ * The entry point to use Datadog's Session Replay feature.
+ */
+class DdSessionReplay(
+    reactContext: ReactApplicationContext
+) : NativeDdSessionReplaySpec(reactContext) {
+
+    private val implementation = DdSessionReplayImplementation()
```
Comment (0xnm) [inline packages/react-native-session-replay/android/src/newarch/kotlin/com/datadog/sessionreplay/reactnative/DdSessionReplay.kt]: why do we need a dedicated implementation class? what is its purpose?

UPD: is it to have a single class which can be used from both old arch and new arch plugins?

File: packages/react-native-session-replay/android/src/test/kotlin/com/datadog/sessionreplay/reactnative/DdSessionReplayTest.kt [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import org.mockito.junit.jupiter.MockitoExtension
+import org.mockito.junit.jupiter.MockitoSettings
+import org.mockito.quality.Strictness
+
+@Extensions(
+    ExtendWith(MockitoExtension::class),
+    ExtendWith(ForgeExtension::class)
+)
+@MockitoSettings(strictness = Strictness.LENIENT)
+internal class DdSessionReplayTest {
```
Comment (0xnm) [inline packages/react-native-session-replay/android/src/test/kotlin/com/datadog/sessionreplay/reactnative/DdSessionReplayTest.kt]: let's call it `DdSessionReplayImplementationTest` to avoid the confusion. Otherwise one may think that class under test is `DdSessionReplay`, and this is not true.

File: packages/react-native-session-replay/android/gradle.properties (hunk 5 above)
Comment (louiszawadzki) [reply packages/react-native-session-replay/android/gradle.properties]: I'd like to keep this in sync with the value used in the version of the RN dev dependency (currently 0.71).
So I agree it should be at least 33 which is the default version for RN 0.71 and 0.72.

Is it ok if I bump it to 34 once we migrate the RN dev dependency to 0.73?

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/sessionreplay/reactnative/DdSessionReplayImplementation.kt (hunk 6 above)
Comment (louiszawadzki) [reply packages/react-native-session-replay/android/src/main/kotlin/com/datadog/sessionreplay/reactnative/DdSessionReplayImplementation.kt]: In this PR I want to make sure that all native packages work (i.e. the app does not crash when they're called) without any actual logic in them (it's mostly "plumbing" work).
In the next PR I'll enable SR in this method.

File: packages/react-native-session-replay/android/src/newarch/kotlin/com/datadog/sessionreplay/reactnative/DdSessionReplay.kt (hunk 7 above)
Comment (louiszawadzki) [reply packages/react-native-session-replay/android/src/newarch/kotlin/com/datadog/sessionreplay/reactnative/DdSessionReplay.kt]: Yes exactly, this is not great but is the current standard when supporting both old and new architecture in RN libs.

File: packages/react-native-session-replay/android/gradle.properties (hunk 5 above)
Comment (0xnm) [reply packages/react-native-session-replay/android/gradle.properties]: My question is just theoretical, let's keep it in sync with the value used by the latest RN version we are using.

File: packages/react-native-session-replay/android/build.gradle [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+    // noinspection GradleDynamicVersion
+    api 'com.facebook.react:react-native:+'
+  } else {
+    // We specify the $reactNativeVersion, like it's done on the react-native-gradle-plugin, as the plugin is not applied in tests.
+    // There is no impact for apps as we apply the same logic:
+    // https://github.com/facebook/react-native/blob/e1a1e6aa8030bf11d691c3dcf7abd13b25175027/packages/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/utils/DependencyUtils.kt
+    api "com.facebook.react:react-android:$reactNativeVersion"
+  }
+  implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
+  compileOnly "com.squareup.okhttp3:okhttp:3.12.13"
```
Comment (0xnm) [inline packages/react-native-session-replay/android/build.gradle]: FYI: SDK v2 is on OkHttp 4.11 https://github.com/DataDog/dd-sdk-android/blob/4744e9d7f85449708c96580ba5f7bd9695dbebc0/gradle/libs.versions.toml#L6, so you probably would like to change the version here.

File: packages/react-native-session-replay/android/src/main/kotlin/com/datadog/sessionreplay/reactnative/DatadogSDKReactNativeSessionReplayPackage.kt
Comment (0xnm) [inline packages/react-native-session-replay/android/src/main/kotlin/com/datadog/sessionreplay/reactnative/DatadogSDKReactNativeSessionReplayPackage.kt:1]: all files should be moved to conform the new namespace, they shouldn't be under `com/datadog/sessionreplay/reactnative`, but under `com/datadog/reactnative/sessionreplay`

File: packages/react-native-session-replay/android/build.gradle (hunk 9 above)
Comment (louiszawadzki) [reply packages/react-native-session-replay/android/build.gradle]: Indeed it's not needed here. 
We only use it for the ProxyAuthenticator and it's `compileOnly`, so the migration should be fairly easy for `core`.
I will create a technical task for the migration.

File: packages/react-native-session-replay/android/build.gradle [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+    release {
+      minifyEnabled false
+    }
+  }
+  lintOptions {
+    disable 'GradleCompatible'
+  }
+  compileOptions {
+    sourceCompatibility JavaVersion.VERSION_1_8
+    targetCompatibility JavaVersion.VERSION_1_8
```
Comment (0xnm) [inline packages/react-native-session-replay/android/build.gradle]: SDK v2 is producing Java 11 bytecode, so maybe you would like to align

File: packages/react-native-session-replay/android/build.gradle [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+  unmock 'org.robolectric:android-all:4.4_r1-robolectric-r2'
+}
+
+unMock {
+  keep("android.os.Looper")
+  keep("android.os.MessageQueue")
+  keep("android.os.SystemProperties")
+  keep("android.view.Choreographer")
+  keep("android.view.DisplayEventReceiver")
+}
```
Comment (0xnm) [inline packages/react-native-session-replay/android/build.gradle]: not sure if all this is still needed with SDK v2

**Review by 0xnm — DISMISSED** [review]

**Comment by louiszawadzki** [comment]
@buraizu I wanted to add an indication in case someone noticed it browsing github and would attempt to use it right away.

File: packages/react-native-session-replay/android/build.gradle (hunk 10 above)
Comment (louiszawadzki) [reply packages/react-native-session-replay/android/build.gradle]: What is the risk with changing this value?
I would prefer to make the change in a dedicated PR to avoid breaking support for some client apps.

File: packages/react-native-session-replay/android/build.gradle (hunk 11 above)
Comment (louiszawadzki) [reply packages/react-native-session-replay/android/build.gradle]: Not needed for RNSR but still needed for core tests. I will remove in the next PR so you don't have to approve this one again for it to be merged :)

File: packages/react-native-session-replay/android/build.gradle (hunk 10 above)
Comment (0xnm) [reply packages/react-native-session-replay/android/build.gradle]: I guess there is no risk, because SDK v2 artifacts are already with Java 11 bytecode anyway. Out of curiosity, what bytecode level is emitted by RN itself?

Comment (louiszawadzki) [reply packages/react-native-session-replay/android/build.gradle]: It's `VERSION_11` for the latest versions.

File: packages/react-native-session-replay/android/build.gradle [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+  testImplementation "org.junit.jupiter:junit-jupiter-params:5.6.2"
+  testImplementation "org.mockito:mockito-junit-jupiter:3.4.6"
+  testImplementation "org.assertj:assertj-core:3.18.1"
+  testImplementation "com.github.xgouchet.Elmyr:core:1.3.1"
+  testImplementation "com.github.xgouchet.Elmyr:inject:1.3.1"
+  testImplementation "com.github.xgouchet.Elmyr:junit5:1.3.1"
+  testImplementation "com.github.xgouchet.Elmyr:jvm:1.3.1"
+  testImplementation "com.nhaarman.mockitokotlin2:mockito-kotlin:2.2.0"
+  testImplementation "org.jetbrains.kotlin:kotlin-reflect:$kotlin_version"
+  unmock 'org.robolectric:android-all:4.4_r1-robolectric-r2'
```
Comment (0xnm) [inline packages/react-native-session-replay/android/build.gradle:183]: probably you don't need this either and no need to apply `unmock` plugin as well, but this is very minor, can be addressed later

**Review by 0xnm — APPROVED** [review]
